R/utils.R

```diff
@@ -259,84 +259,333 @@ rowcol_to_dim <- function(row, col) {
+
+  rows_arg
+  cols_arg
+
+  if (!all(length(scol) == 1, length(srow) == 1, scol >= 0, srow >= 0)) {
+    stop("Internal error. At this point scol and srow should have length 1.")
+  }
+
+
+  # if `!x` return early
+  if (!x_present) {
+    row_span <- srow + rows_arg %||% 1L
+    col_span <- scol + cols_arg %||% 1L
+    if (length(row_span) == 1 && length(col_span) == 1) {
+      # A1
+      row_start <- row_span
+      col_start <- col_span
+      dims <- rowcol_to_dim(row_start, col_start)
+    } else {
+      # A1:B2
+
+      dims <- rowcol_to_dims(row_span, col_span)
+    }
+    return(dims)
+  }
+  # Making sure that at this point, we only cover the case for `x`
+
+  col_names <- col_names %||% x_has_named_dims
+  row_names <- row_names %||% FALSE
+  if (!col_names && row_names && x_has_named_dims) {
+    warning("The combination of `row_names = TRUE` and `col_names = FALSE` is not recommended.",
+      "`col_names` allows to select the region that contains the data only.",
+      "`row_names` = TRUE adds row numbers if the data doesn't have rownames.",
+      call. = FALSE
+    )
+  }
+  assert_class(col_names, "logical")
+  assert_class(row_names, "logical")
+  x <- as.data.frame(x)
+
+  nrow_to_span <- nrow(x)
+  ncol_to_span <- ncol(x)
+
+  if (col_names && x_has_named_dims) {
+    nrow_to_span <- nrow_to_span + 1
+  }
+  # if without column names and with named dimensions
+  # We will increment the start row by 1.
+  if (!col_names && x_has_named_dims) {
+    srow <- srow + 1
+  }
+
+  # Adding a column when spanning.
+  if (row_names) {
+    ncol_to_span <- ncol_to_span + 1
+  }
+
+  # if (row_names) {
+  #   scol <- scol + 1
+  # }
+
+  if (!all(scol >= 0, srow >= 0)) {
+    stop("Internal error. At this point `start_col` and `start_row` should have length 1.")
+  }
+
+  if (is.null(cols_arg) && is.null(rows_arg)) {
+    # wb_dims(data.frame())
+    row_span <- srow + seq_len(nrow_to_span)
+    col_span <- scol + seq_len(ncol_to_span)
+  } else if (is.null(rows_arg)) {
+    row_span <- srow + seq_len(nrow_to_span)
+    col_span <- scol + cols_arg + row_names
+  } else if (is.null(cols_arg)) {
+    row_span <- srow + rows_arg + col_names
+    col_span <- scol + seq_len(ncol_to_span)
+  } else {
+    "problem"
+  }
+
+  if (length(row_span) == 1 && length(col_span) == 1) {
     # A1
-    dims <- rowcol_to_dim(rows, cols)
+    row_start <- row_span
+    col_start <- col_span
+    dims <- rowcol_to_dim(row_start, col_start)
   } else {
     # A1:B2
-    dims <- rowcol_to_dims(rows, cols)
+
+    dims <- rowcol_to_dims(row_span, col_span)
   }
 
   dims
 }
 
+# It is inspired heavily by `rlang::arg_match(multi = TRUE)` and `base::match.arg()`
+match.arg_wrapper <- function(arg, choices, several.ok = FALSE, fn_name = NULL) {
+  # Check valid argument names
+  # partial matching accepted
+  fn_name <- fn_name %||% "function"
+  # match.arg(arg, choices = choices, several.ok = several.ok)
+  # Using rlang::arg_match() would remove that.
```

Comment:
Just checked, `rlang` apparently only has `utils` as dependency, but unless we really really really ... really cannot avoid this, I don't want any new dependencies.

Reply:
If you think this function is correct, let's not add it. ;)